File: 2023/day2.py
import math
import re
from collections import Counter

from util import parse_file_line_by_line


def parse_line(line: str):
    if captures := re.match(r"Game (\d+): (.+)", line):
        id_, game = captures.groups()
        colors: Counter[str] = Counter()
        for amount, color in re.findall(r"(\d+) (\w+)", game):
            colors[color] = max(colors[color], int(amount))
        return int(id_), colors

def part_one(line: str) -> int:
    max_colors = Counter(red=12, green=13, blue=14)
    id_, colors = parse_line(line)
    return id_ if colors <= max_colors else 0


def part_two(line: str) -> int:
    _, colors = parse_line(line)
    return math.prod(colors.values())

if __name__ == "__main__":
    print(sum(parse_file_line_by_line(2, 1, part_one)))
    print(sum(parse_file_line_by_line(2, 2, part_two)))

File: 2023/Day2.java
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Stream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;

class Day2 {
    private static final Map<String, Integer> maxCubes = Map.of("red", 12,"green", 13, "blue", 14);
    public static void main(String[] args) throws IOException{
        Path path = Paths.get("inputs/2.txt");
        Stream<String> lines = Files.newBufferedReader(path).lines();
        Integer part1 = lines.map(Day2::partOne).mapToInt(l -> l).sum();
        System.out.println(part1);

        lines = Files.newBufferedReader(path).lines();
        Integer part2 = lines.map(Day2::partTwo).mapToInt(l -> l).sum();
        System.out.println(part2);

    }

    private static HashMap<String, Integer> partOneAndTwoHelper(String line) {
        Pattern patternGame = Pattern.compile("Game (\\d+): (.+)");
        Matcher matcherGame = patternGame.matcher(line);
        Integer gameID = Integer.valueOf(0);
        HashMap<String, Integer> cubeCounts = new HashMap<>();
        cubeCounts.put("red", 0);
        cubeCounts.put("green", 0);
        cubeCounts.put("blue", 0);

        if (matcherGame.find()) {
            gameID = Integer.valueOf(matcherGame.group(1)); 
            cubeCounts.put("game", gameID);
            Pattern patternSet = Pattern.compile("(\\d+) (\\w+)");
            Matcher matcherSet = patternSet.matcher(line);
            while (matcherSet.find()){
                Integer amount = Integer.valueOf(matcherSet.group(1));
                String color = matcherSet.group(2);
                cubeCounts.put(color, Math.max(amount, cubeCounts.get(color)));
            }
        };
        return cubeCounts;
    }

    private static Integer partOne(String line){
        HashMap<String, Integer> cubeCounts= partOneAndTwoHelper(line);
        
        for (String key : maxCubes.keySet()) {
            if (cubeCounts.get(key) > maxCubes.get(key)) {
                return Integer.valueOf(0);
            }
        }
        return cubeCounts.get("game");
    }

    private static Integer partTwo(String line){
        HashMap<String, Integer> cubeCounts= partOneAndTwoHelper(line);
        return cubeCounts.get("blue") * cubeCounts.get("red") * cubeCounts.get("green");
    }
}